go-controller/pkg/clustermanager/userdefinednetwork/controller.go

```diff
@@ -0,0 +1,361 @@
+					return nil, fmt.Errorf("failed to verify NAD not in use [%s/%s]: %w", nad.Namespace, nad.Name, &networkInUseError{err: err})
+				}
+
+				controllerutil.RemoveFinalizer(nad, template.FinalizerUserDefinedNetwork)
+				nad, err := c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(nad.Namespace).Update(context.Background(), nad, metav1.UpdateOptions{})
+				if err != nil {
+					return nil, err
+				}
+				klog.Infof("Finalizer removed from NetworkAttachmetDefinition [%s/%s]", nad.Namespace, nad.Name)
+			}
+
+			controllerutil.RemoveFinalizer(udn, template.FinalizerUserDefinedNetwork)
+			udn, err := c.udnClient.K8sV1().UserDefinedNetworks(udn.Namespace).Update(context.Background(), udn, metav1.UpdateOptions{})
+			if err != nil {
+				return nil, fmt.Errorf("failed to remove finalizer to UserDefinedNetwork: %w", err)
+			}
+			klog.Infof("Finalizer removed from UserDefinedNetworks [%s/%s]", udn.Namespace, udn.Name)
+		}
+
+		return nad, nil
+	}
+
+	if finalizerAdded := controllerutil.AddFinalizer(udn, template.FinalizerUserDefinedNetwork); finalizerAdded {
+		udn, err := c.udnClient.K8sV1().UserDefinedNetworks(udn.Namespace).Update(context.Background(), udn, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to add finalizer to UserDefinedNetwork: %w", err)
+		}
+		klog.Infof("Added Finalizer to UserDefinedNetwork [%s/%s]", udn.Namespace, udn.Name)
+	}
+
+	desiredNAD, err := c.renderNadFn(udn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate NetworkAttachmentDefinition: %w", err)
+	}
+	if nad == nil {
+		// creating NAD in case no primary network exist should be atomic and synchronized with
+		// any other thread that create NADs.
+		// Since the UserDefinedNetwork controller use single thread (threadiness=1),
+		// and being the only controller that create NADs, this conditions is fulfilled.
+		if primaryNetwork(udn.Spec) {
+			actualNads, lerr := c.nadLister.NetworkAttachmentDefinitions(udn.Namespace).List(labels.Everything())
+			if lerr != nil {
+				return nil, fmt.Errorf("failed to list  NetworkAttachmetDefinition: %w", lerr)
+			}
+			if err := validatePrimaryNetworkNADNotExist(actualNads); err != nil {
+				return nil, err
+			}
+		}
+
+		nad, err = c.nadClient.K8sCniCncfIoV1().NetworkAttachmentDefinitions(udn.Namespace).Create(context.Background(), desiredNAD, metav1.CreateOptions{})
+		if err != nil {
+			return nil, fmt.Errorf("failed to create NetworkAttachmentDefinition: %w", err)
+		}
+		klog.Infof("Created NetworkAttachmentDefinition [%s/%s]", nad.Namespace, nad.Name)
+		return nad, nil
+	}
+
+	if !metav1.IsControlledBy(nad, udn) {
+		return nil, fmt.Errorf("foreign NetworkAttachmentDefinition with the desired name already exist [%s/%s]", nad.Namespace, nad.Name)
+	}
+
+	if !reflect.DeepEqual(nad.Spec, desiredNAD.Spec) {
+		nad.Spec.Config = desiredNAD.Spec.Config
+		nad.ObjectMeta.OwnerReferences = desiredNAD.ObjectMeta.OwnerReferences
```

Comment:
should we remove this line if we are not tracking OwnerRef changes?

Reply:
Done